data/basal.go

```diff
@@ -0,0 +1,141 @@
+package data
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	valid "github.com/tidepool-org/platform/Godeps/_workspace/src/gopkg.in/bluesuncorp/validator.v8"
+
+	"github.com/tidepool-org/platform/validate"
+)
+
+func init() {
+	validator.RegisterValidation(rateTag, BasalRateValidator)
+	validator.RegisterValidation(durationTag, BasalDurationValidator)
+	validator.RegisterValidation(deliveryTypeTag, BasalDeliveryTypeValidator)
+	validator.RegisterValidation(insulinTag, BasalInsulinValidator)
+	validator.RegisterValidation(valueTag, BasalValueValidator)
+}
+
+type Basal struct {
```

Comment:
A couple comments about data types. 

After working on the de-duplication problem and the variability of raw device data (and the potential for variability of Tidepool data, particularly when 3rd parties start uploading data), I now think we should be as strict as possible. (Screw being lenient, we need to make sure our data is as pristine as possible. Bad data is a huge tech debt to deal with.)  I think it does make sense to be lenient in some areas of our API, but not the data.  For example, we could be lenient on signup when the user enters an email address (we should accept any case, but force it to lowercase before creating the account).

Also, I think we should be strict not only about the format of an individual fields, but also how the entire data point looks.  For example, for a Basal data point, many of these fields should be absolutely required.  Also, we should enforce extra validations upon data fields are interdependent (for example, if we allowed both 'mg/dL' and 'mmol' units then the range of acceptable values should be tied to the units).

My 2 cents... I guess I'm just getting paranoid about our data.  After all, our data is everything. Without clean data our usefulness is reduced and dirty data is just a huge tech debt.  This was reinforced with all the data de-dup work.